extendr-macros/src/wrappers.rs

```diff
@@ -120,6 +145,43 @@ pub fn make_function_wrappers(
             });)
         })
         .unwrap_or_default();
+
+    // figure out if &Self / &mut Self / inside of impl block &ImplType / &mut ImplType is used!
+    let return_is_ref_self = {
+        match sig.output {
+            syn::ReturnType::Default => false,
+            syn::ReturnType::Type(_, ref return_type) => match return_type.as_ref() {
+                // matches `-> Self`
+                // Type::Path(type_path) => type_path.path.is_ident("Self"),
+                // matches `-> &Self` / `-> &mut Self`
+                Type::Reference(ref reference_type) => {
+                    if let Type::Path(path) = reference_type.elem.as_ref() {
+                        path.path.is_ident("Self")
+                            || self_ty
+                                .map(|x| x == reference_type.elem.as_ref())
+                                .unwrap_or(false)
+                    } else {
+                        false
+                    }
+                }
+                _ => false,
+            },
+        }
+    };
+
+    let return_type_conversion = if !return_is_ref_self {
+        quote!(Ok(extendr_api::Robj::from(#call_name(#actual_args))))
+    } else {
+        // instead of converting &Self / &mut Self, pass on the passed
+        // ExternalPtr<Self>
+        quote!(
+            let _return_ref_to_self = #call_name(#actual_args);
+            //FIXME: find a less hardcoded way to write `_self_robj`
+            Ok(_self_robj)
+        )
+    };
```

Comment:
I suggest flipping the order of the if else so that it is more readable instead of negating the bool value

```suggestion
    let return_type_conversion = if return_is_ref_self {
         // instead of converting &Self / &mut Self, pass on the passed
        // ExternalPtr<Self>
        quote!(
            let _return_ref_to_self = #call_name(#actual_args);
            //FIXME: find a less hardcoded way to write `_self_robj`
            Ok(_self_robj)
        )
    } else {
      quote!(Ok(extendr_api::Robj::from(#call_name(#actual_args))))
    };
```